File: cpp/cpp_100-999/513.cc
// C++
// leetcode 513
// https://leetcode.cn/problems/find-bottom-left-tree-value/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int findBottomLeftValue(TreeNode* root) {
    	queue<TreeNode*> q;
    	q.push(root);
    	int level = 0, res = 0;
    	while(!q.empty()) {
    		int size = q.size();
    		res = q.front()->val;
    		while(size--) {
    			auto p = q.front();
    			if(p->left)
    				q.push(p->left);
    			if(p->right)
    				q.push(p->right);
    			q.pop();
    		}
    	}
    	return res;
    }
};

File: Go/Go_100-999/513.go
// Go
// leetcode 513
// https://leetcode.cn/problems/find-bottom-left-tree-value/

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func findBottomLeftValue(root *TreeNode) (res int) {
	q := []*TreeNode{root}
	for len(q) != 0 {
		res = q[0].Val
		size := len(q)
		for i := 0; i < size; i++ {
			node := q[i]
			if node.Left != nil {
				q = append(q, node.Left)
			}
			if node.Right != nil {
				q = append(q, node.Right)
			}
		}
		q = q[size:]
	}
	return res
}